jbmc/src/java_bytecode/README.md

```diff
@@ -39,9 +39,160 @@ To be documented.
 
 To be documented.
 
-\section java-bytecode-remove-exceptions remove_exceptions
+\section java-bytecode-remove-exceptions Remove exceptions
 
-To be documented.
+When `remove_exceptions` is called on the goto model, the goto model contains
+complex instructions such as `CATCH-POP`, `CATCH-PUSH` and `THROW`. In order to
+analyze the goto model, the instructions must be simplified to use more basic
+instructions - this is called "lowering". This class lowers the `CATCH` and
+`THROW` instructions.
+
+Exceptions that have been thrown, but not yet caught are stored the global
+`@inflight_exception`. When they have been caught, the global
+`@inflight_exception` is set back to null and ready for the next exception.
+
+\subsection Throw
+
+Consider a simple method `testException(I)V`:
+
+```java
+public void testException(int i) throws Exception {
+    if (i < 0) {
+        throw new Exception();
+    }
+    field = i;
+}
+```
+
+The goto for `testException(I)V` before `remove_exceptions` (removing comments
+and replacing irrelevant parts with `...`) is:
+
+```
+com.diffblue.regression.TestExceptions.testException(int) /* java::com.diffblue.regression.TestExceptions.testException:(I)V */
+  IF i >= 0 THEN GOTO 3
+  struct java.lang.Exception *new_tmp0;
+  new_tmp0 = new struct java.lang.Exception;
+  IF !(new_tmp0 == null) THEN GOTO 1
+
+...
+
+1: SKIP
+  new_tmp0 . java.lang.Exception.<init>:()V();
+
+...
+
+2: SKIP
+  THROW: throw(new_tmp0)
+  dead new_tmp0;
+
+...
+
+```
+where there is a `THROW` instruction to be replaced.
+
+After passing the goto model through `remove_exceptions`, it is:
+
+```
+com.diffblue.regression.TestExceptions.testException(int) /* java::com.diffblue.regression.TestExceptions.testException:(I)V */
+  IF i >= 0 THEN GOTO 4
+  struct java.lang.Exception *new_tmp0;
+  new_tmp0 = new struct java.lang.Exception;
+  IF !(new_tmp0 == null) THEN GOTO 1
+
+...
+
+1: new_tmp0 . java.lang.Exception.<init>:()V();
+  IF @inflight_exception == null THEN GOTO 2 // it is because we've not used it yet
+
+...
+
+2: IF !(new_tmp0 == null) THEN GOTO 3
+
+...
+
+3: @inflight_exception = (void *)new_tmp0;
+  dead new_tmp0;
+
+...
+
+```
+where now instead of the instruction `THROW`, the global variable
+`@inflight_exception` holds the thrown exception in ia separate goto statement.
+
+\subsection Catch
+
+Consider the method `catchSomething(I)V` that tries the above method `testException(I)V` and catches the exception:
+
+```java
+public void catchSomething(int i) {
+    try {
+        testException(i);
+    } catch (Exception e) {
+    }
+}
+```
+
+The goto model before `remove_exceptions` is:
+```
+com.diffblue.regression.TestExceptions.catchSomething(int) /* java::com.diffblue.regression.TestExceptions.catchSomething:(I)V */
+
+...
+
+  CATCH-PUSH ->2
```

Comment:
This needs some context I think?

Reply:
There is not much context to give, I'm afraid.